### ISSUE TITLE
Query: potential bug in null semantics VisitSqlBinary 

### ISSUE THREAD

**Issue by maumar** [dotnet/efcore#30247]
WIP

**Comment by ranma42** [comment]
@maumar IIUC in the case you describe the analysis is returning correct (inaccurate) nullability information; instead of reporting the more precise result `(value: constant C, nullable: false)` it returns `(value: constant C, nullable: true)`.

This should likely not be much of a problem as the result is still sound.
In order to make the query pipeline more *robust* it should be able to work just fine with everything marked as nullable (this would regress the queries, but ideally should not break anything).
OTOH propagating the updated nullability information might make it possible to have more efficient queries.

Note that if the nullability analysis determined (incorrectly) that an expression is non-nullable and then the actual query returned `NULL`s, then you would have bugs (caused by an unsound analysis, i.e. an analysis that "proved" something false).

**Comment by maumar** [comment]
I agree, there shouldn't be risk of functional errors (i.e. data corruption) but potential missed optimizations. I'm not too concerned about this, but filed an issue to track this "just in case" thinking we may want to address it "in the future". But I agree its a very low priority issue.
